lib/modules/auth/providers/auth.provider.dart

```diff
@@ -1,9 +1,40 @@
+import 'package:dio/dio.dart';
 import 'package:flutter/src/widgets/framework.dart';
+import 'package:project/config/inject.config.dart';
 import 'package:project/core/app_provider.dart';
+import 'package:project/core/exceptions/unauthorized.dart';
+import 'package:project/modules/auth/services/auth.service.dart';
 
 class AuthProvider extends AppProvider {
   AuthProvider(BuildContext context) : super(context);
 
-  // TODO - Implement
-  // This should be similar to the Mobile Applications provider.
+  Dio dioInstance = inject.get<Dio>();
+
+  String? authToken = null;
+  bool loading = false;
+  String? error = null;
+  AuthService authService = inject.get<AuthService>();
+
+  Future<void> login(String email, String password) {
+    loading = true;
+    error = null;
+    notify('login', notificationType: NotificationType.Start);
+
+    return authService.login(email, password).then((res) {
+      //here we will set the auth token
+      authToken = 'authToken';
+      loading = false;
+      notify('login', notificationType: NotificationType.Success);
+    }).catchError((err) {
+      switch (err.runtimeType) {
+        case (UnauthorizedException):
+          error = 'Wrong email or password';
+          break;
+        default:
+          error = 'Oops. An error occurred';
+      }
```

Comment:
Incerc sa dau un exemplu mai concret de cum imi imaginez ca ar trebui sa ar trebui sa facem error handling.

**Service:** prinde DioError si arunca un Exception custom de-al nostru
**Provider:** prinde Exception custom din service si il salveaza in state fara modificari
**Widget:** citeste Exception custom din state si arata mesaj in functie de ce tip e

Toata structura asta vine din faptul ca providerul ar trebui doar sa tina si sa modifice state-ul, fara orice alta logica.
Putem sa folosim si structura asta pe care ai scris-o dar o sa ne lovim de limitari cand incepem sa implementam traducerile.

Also nu esti limitat sa folosesti .then().catchError(). Poti linistit sa faci functiile din provider async si sa astepti rezultatul service-ului cu await intr-un try catch daca crezi ca ii mai usor de scris si inteles.

Reply:
Personal prefer then si catchAll deci ii ok pt mine. Inteleg ca trebuie teoretic sa luam in considerare cat mai multe, dar suntem siguri ca o sa ajungem sa facem traduceri or smthn like that? Trebuie gasit un echilibru intre quality si quantity pentru ca altfel riscam sa facem chestiile bine, dar sa nu apucam sa facem mare lucru si asta ii mai rau decat sa faci mult, dar cu quality un pic mai prost. (in contextul materiei)